### PR TITLE
service: Handle backend with initial state set to Terminating

### PR DESCRIPTION
**Pull request by sterchelen**
Backends were previously set by default with BackendStateActive as
initial state. Even though it seems strange you can have endpointslices
starting in a terminating state.

When the EnableK8sTerminatingEndpoint knob is activated in
pkg/k8s/endpoints.go:ParseEndpointSlice, endpoints in terminating state
aren't skipped. Now moving forward in the logic, in UpsertService
we construct the backends list without considering its initial state.
With these two conditions, a new backend even in BackendTerminatingState
was considered active b/c of the default BackendStateActive and thus
mechanically added to the bpf backends map.
This was resulting in packets sent to a remote not accepting any
connections.

This patch fixes this situation by not setting a default state to the
backend.

---
### Here the proof that endpointslice in terminating mode (as initial state) was considered active

Below the output of a homemade tool subscribed to a specific endpointslice's events.
1. A new endpoint `chazam-78bcf8785c-769zv` added with ready, serving & terminating set to false.
2. Without waiting its readiness I deleted it right after which resulted in an endpoint with ready, serving set to false **but** terminating set to **true**
```csv
action, ip, pod_name, node_name, ready, serving, terminating
newEndpoint,100.97.232.213,chazam-78bcf8785c-769zv,i-0e9b6f012192efa19,false,false,false      
updateCondition,100.97.232.213,chazam-78bcf8785c-769zv,i-0e9b6f012192efa19,false,false,true
```

Now below the output of `cilium service list`, **before** `kubectl delete pod chazam-78bcf8785c-769zv`
```
36    100.68.110.241:80      ClusterIP      1 => 100.96.90.103:80 (active)                                                                                                                                                       
                                            2 => 100.98.105.156:80 (active)                                                                                                                                                      
                                            3 => 100.96.251.251:80 (active)                                                                                                                                                      
                                            4 => 100.98.119.199:80 (active)                                                                                                                                                      
                                            5 => 100.98.127.136:80 (active)                                                                                                                                                      
                                            6 => 100.96.153.171:80 (active)                                                                                                                                                      
                                            7 => 100.98.134.176:80 (active)                                                                                                                                                      
                                            8 => 100.98.135.206:80 (active)                                                                                                                                                      
                                            9 => 100.98.143.190:80 (active)                                                                                                                                                      
                                            10 => 100.98.144.191:80 (active)
```

Now, after executing `kubectl delete pod chazam-78bcf8785c-769zv` we can see that the `100.97.232.213` ip was added as an active backend.

```
36    100.68.110.241:80      ClusterIP      1 => 100.96.90.103:80 (active)                                                                                                                                                       
                                            2 => 100.98.105.156:80 (active)                                                                                                                                                      
                                            3 => 100.96.251.251:80 (active)                                                                                                                                                      
                                            4 => 100.98.119.199:80 (active)                                                                                                                                                      
                                            5 => 100.98.127.136:80 (active)                                                                                                                                                      
                                            6 => 100.96.153.171:80 (active)                                                                                                                                                      
                                            7 => 100.98.134.176:80 (active)                                                                                                                                                      
                                            8 => 100.98.135.206:80 (active)                                                                                                                                                      
                                            9 => 100.98.143.190:80 (active)                                     
                                            10 => 100.98.144.191:80 (active)                                    
                        --->              11 => 100.97.232.213:80 (active)
```
---
This issue is gone in v1.13 b/c of this specific commit https://github.com/cilium/cilium/commit/ed118b430a70d50bd5cd3fbfeb2f7f472e77e3f9#diff-2b655ba613ca8386fd9f4c383200480492ee25711e24a575f3fc8b33531b4672L1468 where basically no more default state is set to backends.

### CODE REVIEW

**Comment by pchaigno** [comment]
@borkmann I'm assigning you as a reviewer given it touches the LB handling area, but feel free to reassign as needed.

**Comment by borkmann** [comment]
@sterchelen could you elaborate on the behavior when the agent gets restarted? Would this cause temporary drop or would they remain in active if they were considered active before?

Cc @aditighag / @oblazek for reviews

**Review by oblazek — APPROVED** [review]
I think this makes perfect sense. LGTM

**Comment by sterchelen** [comment]
> Could you elaborate on the behavior when the agent gets restarted?

@borkmann Didn't test it as this scenario wasn't part of the observed behavior. If it's needed, I could make a try.

**Comment by sterchelen** [comment]
hey guys, we would like to deploy this patch in our clusters, possible to have a review 🙏🏼?

**Review by aditighag — CHANGES_REQUESTED** [review]
>Backends were previously set by default with BackendStateActive as
initial state. Even though it seems strange you can have endpointslices
starting in a terminating state.

Can you explain when you can have endpoint slices starting in a terminating state?

**Comment by sterchelen** [comment]
> Can you explain when you can have endpoint slices starting in a terminating state?

Sure @aditighag, I guess I need to rephrase a bit. 

From the endpointslice pov (k8s state), **the endpoint isn't starting in a terminating state you are right** and thus, at this moment, it is not added to the cilium's internal state.
Though when the endpoint is marked as terminating (in k8s state) right after being created and with the conjunction of `EnableK8sTerminatingEndpoint` knob, the endpoint starts being handled by cilium and is marked as `BackendStateActive` b/c we don't consider its k8s state while being translated to cilium struct.

---

Now in our context, we have [karpenter](karpenter.sh/) installed in our cluster, we are experiencing this transition from created > terminating state when karpenter consolidates too quickly our instances:
* vm created, configured
* cilium scheduled & running
* business pod A scheduled > business pod A evicted --> **this transition, when a pod isn't transitioning to a ready state, generates the issue**

I hope to have answered your question 🙏🏼.

**Review by aditighag — APPROVED** [review]
Thanks for the context @sterchelen. The fix LGTM for the corner case you mentioned.
I've approved and triggered the test runs, once they pass, we can mark it as ready for merge accordingly.

**Comment by aditighag** [comment]
The unit test added in the PR passed, marking it as ready for merge.

**Comment by joamaki** [comment]
/test-1.12

**Comment by joamaki** [comment]
> /test

The fix was against 1.12, hence why e.g. ci-l4lb failed with `Error: unknown flag: --backend-weights`.

**Comment by squeed** [comment]
Can you rebase on latest master? Then the failing tests should pass.

**Comment by squeed** [comment]
Kicked a failed to-external test.

**Review by squeed — APPROVED** [review]

**Comment by julianwiedmann** [comment]
/ci-l4lb-1.12

**Pull request merged**
